### PR TITLE
Fix usage of fs.listdir in CheckpointConnector

### LINKED ISSUE #15379: URL passed to fs.listdir instead of a path
### Bug description

https://github.com/Lightning-AI/lightning/blob/master/src/pytorch_lightning/trainer/connectors/checkpoint_connector.py#L582
This line is called `fs.listdir(dir_path)` where `dir_path` is an URL instead of a path.

pyarrow will complain about it:
```
pyarrow.lib.ArrowInvalid: FileSelector.base_dir must not be a URI, got: hdfs:///somewhere/MNIST
```

### How to reproduce the bug

use a HDFS path in trainer's default_root_dir

### Error messages and logs

```
  File "main.py", line 63, in main
    trainer.fit(mnist_model, train_loader, valid_loader)
  File "/home/jobuser/build/yuxlu-test/environments/satellites/python/lib/python3.10/site-packages/pytorch_lightning/trainer/trainer.py", line 696, in fit
    self._call_and_handle_interrupt(
  File "/home/jobuser/build/yuxlu-test/environments/satellites/python/lib/python3.10/site-packages/pytorch_lightning/trainer/trainer.py", line 650, in _call_and_handle_interrupt
    return trainer_fn(*args, **kwargs)
  File "/home/jobuser/build/yuxlu-test/environments/satellites/python/lib/python3.10/site-packages/pytorch_lightning/trainer/trainer.py", line 735, in _fit_impl
    results = self._run(model, ckpt_path=self.ckpt_path)
  File "/home/jobuser/build/yuxlu-test/environments/satellites/python/lib/python3.10/site-packages/pytorch_lightning/trainer/trainer.py", line 1110, in _run
    self._restore_modules_and_callbacks(ckpt_path)
  File "/home/jobuser/build/yuxlu-test/environments/satellites/python/lib/python3.10/site-packages/pytorch_lightning/trainer/trainer.py", line 1063, in _restore_modules_and_callbacks
    self._checkpoint_connector.resume_start(checkpoint_path)
  File "/home/jobuser/build/yuxlu-test/environments/satellites/python/lib/python3.10/site-packages/pytorch_lightning/trainer/connectors/checkpoint_connector.py", line 78, in resume_start
    self.resume_checkpoint_path = self._hpc_resume_path or checkpoint_path
  File "/home/jobuser/build/yuxlu-test/environments/satellites/python/lib/python3.10/site-packages/pytorch_lightning/trainer/connectors/checkpoint_connector.py", line 66, in _hpc_resume_path
    max_version = self.__max_ckpt_version_in_folder(dir_path_hpc, "hpc_ckpt_")
  File "/home/jobuser/build/yuxlu-test/environments/satellites/python/lib/python3.10/site-packages/pytorch_lightning/trainer/connectors/checkpoint_connector.py", line 506, in __max_ckpt_version_in_folder
    files = [os.path.basename(f["name"]) for f in fs.listdir(dir_path)]
  File "/home/jobuser/build/yuxlu-test/environments/satellites/python/lib/python3.10/site-packages/fsspec/spec.py", line 1301, in listdir
    return self.ls(path, detail=detail, **kwargs)
  File "/home/jobuser/build/yuxlu-test/environments/satellites/python/lib/python3.10/site-packages/fsspec/implementations/arrow.py", line 66, in ls
    for entry in self.fs.get_file_info(FileSelector(path))
  File "pyarrow/_fs.pyx", line 433, in pyarrow._fs.FileSystem.get_file_info
  File "pyarrow/error.pxi", line 144, in pyarrow.lib.pyarrow_internal_check_status
  File "pyarrow/error.pxi", line 100, in pyarrow.lib.check_status
pyarrow.lib.ArrowInvalid: FileSelector.base_dir must not be a URI, got: hdfs:///somewhere/MNIST
```

### Environment

fsspec: 2022.10.0
pyarrow: 8.0.0
pytorch_lightning: 1.7.7

### More info

_No response_

cc @awaelchli @ananthsub @ninginthecloud @rohitgr7 @otaj

### PR DESCRIPTION
**Pull request by leoleoasd**
## What does this PR do?

Fix use path in fs.listdir in checkpoint

Fixes #15379

### Does your PR introduce any breaking changes? If yes, please list them.

no

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by leoleoasd** [comment]
Only fixing `listdir` for now. Other usages of `fs.xxxx` still need to be discussed

File: src/pytorch_lightning/trainer/connectors/checkpoint_connector.py [hunk 1]
```diff
@@ -574,12 +575,12 @@ def __max_ckpt_version_in_folder(dir_path: _PATH, name_key: str = "ckpt_") -> Op
         """
 
         # check directory existence
-        fs = get_filesystem(dir_path)
+        fs, uri = url_to_fs(dir_path)
```
Comment (awaelchli) [inline src/pytorch_lightning/trainer/connectors/checkpoint_connector.py:578]: @leoleoasd why do we have to `fs.listdir(uri)` but` fs.exists(dir_path)` is ok?
There are many other places in our code base where we use fsspec (modelcheckpiont, saving utilities, tensorboard, profiler, etc.). Does that mean we need to update the usage there too?

Comment (leoleoasd) [reply src/pytorch_lightning/trainer/connectors/checkpoint_connector.py:578]: https://filesystem-spec.readthedocs.io/en/latest/api.html#fsspec.spec.AbstractFileSystem
According to the document, yes. 
But for mysterious reasons, `exists` works but `listdir` don't work. I'm only changing `listdir` because it blocks my work, and I'm not sure if we have to follow the standard behavior or "if it works, don't touch it". Do we have unit tests for fsspec file systems?

In the document, all functions that aren't bound to a specific FS have parameters named `url` while all methods of `AbstractFileSystem` have parameters named `path`. I think we should update all usages (but not for tensorboard, since tensorboard has its own way of calling fsspec)

**Comment by leoleoasd** [comment]
Hi @Borda @awaelchli any update on the review of this PR?

**Review by carmocca — COMMENTED** [review]
There are several failing tests.

Also, a CHANGELOG entry should be added.

**Comment by leoleoasd** [comment]
Do we actually have any tests for remote file systems? Did anyone try running lightning apps locally with `file://` or `local://`? I don't see any tests in #3805 and #2164.

image
There aren't any string that contains protocol other than HTTP or HTTPS in all PyTorch tests (apart from what I just added).

I added a test that uses fsspec's in-memory file system, and other things related to the test (_hpc_resume_path) are broken too.
I think we DO need to change EVERY usage of fsspec.

File: tests/tests_pytorch/trainer/connectors/test_checkpoint_connector.py [hunk 2]
```diff
@@ -102,6 +102,28 @@ def test_hpc_max_ckpt_version(tmpdir):
     )
 
 
+def test_ckpt_for_fsspec():
+    """Test that the CheckpointConnector is able to write to fsspec file systems."""
+
+    model = BoringModel()
+    # hardcoding dir since `tmpdir` can be windows path
+    trainer = Trainer(default_root_dir="memory://test_ckpt_for_fsspec", max_steps=1)
+    trainer.fit(model)
```
Comment (awaelchli) [inline tests/tests_pytorch/trainer/connectors/test_checkpoint_connector.py:111]:
```suggestion
    trainer = Trainer(default_root_dir="memory://test_ckpt_for_fsspec", limit_train_batches=1, limit_val_batches=1)
    trainer.fit(model)
```

**Review by awaelchli — APPROVED** [review]

**Comment by awaelchli** [comment]
@leoleoasd Responding to your concerns. Yes, we want to support the remote filesystems. It's been added in some places of the code base, logger outputs and checkpointing are the main areas where this is supported. There was never a huge demand, but the users who do use it are mostly using the s3 protocol (from general feedback on slack). For your use cases, you did the right thing and request it on GitHub. Thanks for providing help in improving the coverage here!

File: tests/tests_pytorch/trainer/connectors/test_checkpoint_connector.py [hunk 3]
```diff
@@ -102,6 +102,28 @@ def test_hpc_max_ckpt_version(tmpdir):
...
+    trainer.save_checkpoint("memory://test_ckpt_for_fsspec/hpc_ckpt.ckpt")
+    trainer.save_checkpoint("memory://test_ckpt_for_fsspec/hpc_ckpt_0.ckpt")
```
Comment (carmocca) [inline tests/tests_pytorch/trainer/connectors/test_checkpoint_connector.py:113]: Windows jobs are not happy about these slashes

Comment (leoleoasd) [reply tests/tests_pytorch/trainer/connectors/test_checkpoint_connector.py:113]: I do that to fix tests on windows. fsspec wants / as sep on all platforms.

Comment (leoleoasd) [reply tests/tests_pytorch/trainer/connectors/test_checkpoint_connector.py:113]: I don't have windows testing environment,so I have to push and run tests on GitHub actions. This change is to make fsspec works on windows (we don't want \ in s3 or hdfs paths), but it seems there is another error. I'll fix that later

**Comment by leoleoasd** [comment]
Tests passed, this PR is ready for the next round of review.
The checkpointing issue is fixed, and I'll open new PR if I found some other things that doesn't work with HDFS and add necessary tests.

File: src/pytorch_lightning/trainer/connectors/checkpoint_connector.py [hunk 4]
```diff
@@ -59,13 +61,16 @@ def __init__(self, trainer: "pl.Trainer", resume_from_checkpoint: Optional[_PATH
...
+        fs, path = url_to_fs(dir_path_hpc)
+        if not fs.isdir(path):
+            return None
         max_version = self.__max_ckpt_version_in_folder(dir_path_hpc, "hpc_ckpt_")
         if max_version is not None:
-            return os.path.join(dir_path_hpc, f"hpc_ckpt_{max_version}.ckpt")
+            if isinstance(fs, LocalFileSystem):
+                return os.path.join(dir_path_hpc, f"hpc_ckpt_{max_version}.ckpt")
+            else:
+                return dir_path_hpc + fs.sep + f"hpc_ckpt_{max_version}.ckpt"
```
Comment (awaelchli) [inline src/pytorch_lightning/trainer/connectors/checkpoint_connector.py:73]: I wonder if we will need a utility function for this in the future (I feel like fsspec should have a "join" function). Let's keep it in mind.

**Review by awaelchli — APPROVED** [review]

File: src/pytorch_lightning/trainer/connectors/checkpoint_connector.py (hunk 4 above)
Comment (leoleoasd) [reply src/pytorch_lightning/trainer/connectors/checkpoint_connector.py:73]: According to fsspec/filesystem_spec#747, they don't have that yet but it's coming soon

**Review by carmocca — APPROVED** [review]
LGTM if CI agrees

**Review by otaj — APPROVED** [review]
